### PR TITLE
[ROX-11301] Add new datastore for node components

### PR DESCRIPTION
**Pull request by md2119**
## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- ~Evaluated and added CHANGELOG entry if required~
- ~Determined and documented upgrade steps~
- ~Documented user facing changes~
If any of these don't apply, please comment below.

### CODE REVIEW

File: central/nodecomponent/datastore/datastore.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	searchPkg "github.com/stackrox/rox/pkg/search"
+)
+
+// DataStore is an intermediary to NodeComponent storage.
+//go:generate mockgen-wrapper
+type DataStore interface {
+	Search(ctx context.Context, q *v1.Query) ([]searchPkg.Result, error)
+	SearchNodeComponents(ctx context.Context, q *v1.Query) ([]*v1.SearchResult, error)
```
Comment (connorgorman) [inline central/nodecomponent/datastore/datastore.go:20]: Do we need this one?

File: central/nodecomponent/datastore/index/index.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+// Indexer provides funtionality to index node components.
+type Indexer interface {
+	AddNodeComponent(components *storage.NodeComponent) error
+	AddNodeComponents(components []*storage.NodeComponent) error
+	Count(q *v1.Query, opts ...blevesearch.SearchOption) (int, error)
+	DeleteNodeComponent(id string) error
+	DeleteNodeComponents(ids []string) error
+	MarkInitialIndexingComplete() error
+	NeedsInitialIndexing() (bool, error)
```
Comment (connorgorman) [inline central/nodecomponent/datastore/index/index.go:18]: I think it's okay to omit the ones that are not relevant in Postgres

File: central/nodecomponent/datastore/search/searcher.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/stackrox/rox/pkg/search/blevesearch"
+	pkgPostgres "github.com/stackrox/rox/pkg/search/scoped/postgres"
+)
+
+// Searcher provides search functionality on existing image components.
+//go:generate mockgen-wrapper
+type Searcher interface {
+	Search(ctx context.Context, query *v1.Query) ([]search.Result, error)
+	Count(ctx context.Context, query *v1.Query) (int, error)
+	SearchNodeComponents(context.Context, *v1.Query) ([]*v1.SearchResult, error)
```
Comment (connorgorman) [inline central/nodecomponent/datastore/search/searcher.go:20]: Don't think this will be used so can be omitted

**Review by connorgorman — CHANGES_REQUESTED** [review]
Small changes. I don't think we should carry forward the v1.SearchResult as we don't actually care about registering this with global search right now. In fact, not sure if we should be registering the other components either?

File: central/nodecomponent/datastore/datastore.go (hunk 1 above)
Comment (md2119) [reply central/nodecomponent/datastore/datastore.go:20]: Yea, graphQL

File: central/nodecomponent/datastore/search/searcher.go (hunk 3 above)
Comment (md2119) [reply central/nodecomponent/datastore/search/searcher.go:20]: we need it for graphQL :)

**Comment by md2119** [comment]
> Small changes. I don't think we should carry forward the v1.SearchResult as we don't actually care about registering this with global search right now. In fact, not sure if we should be registering the other components either?

@connorgorman, do we know for sure that our global search is not used directly from APIs? If not, we will have to release a notice about the change. But that also means that we cannot remove it RN.

File: central/nodecomponent/datastore/index/index.go (hunk 2 above)
Comment (md2119) [reply central/nodecomponent/datastore/index/index.go:18]: Fair enough. I seemed have to have copied the common interface pattern between dackbox and postgres.

**Review by connorgorman — APPROVED** [review]

File: central/nodecomponent/datastore/search/searcher.go (hunk 3 above)
Comment (md2119) [reply central/nodecomponent/datastore/search/searcher.go:20]: discussed offline

**Pull request merged**